src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTask.cs

```diff
@@ -0,0 +1,535 @@
+                    var sourcePath = GetSourcePath(packageFile, customMetadata);
+                    if (excludeFiles.Contains(sourcePath))
+                    {
+                        continue;
+                    }
+
+                    GetTargetPath(packageFile, customMetadata, sourcePath, currentProjectDirectory, out targetPaths);
+
+                    if (fileModel.ContainsKey(sourcePath))
+                    {
+                        var setOfTargetPaths = fileModel[sourcePath];
+                        setOfTargetPaths.AddRange(targetPaths);
+                    }
+                    else
+                    {
+                        var setOfTargetPaths = new HashSet<string>();
+                        setOfTargetPaths.AddRange(targetPaths);
+                        fileModel.Add(sourcePath, setOfTargetPaths);
+                    }
+                }
+            }
+
+            return fileModel;
+        }
+
+
+        // The targetpaths returned from this function contain the directory in the nuget package where the file would go to. The filename is added later on to the target path.
+        private void GetTargetPath(ITaskItem packageFile, IDictionary customMetadata, string sourcePath, string currentProjectDirectory, out string[] targetPaths)
+        {
+            targetPaths = new string[] { PackagingConstants.Folders.Content + Path.DirectorySeparatorChar, PackagingConstants.Folders.ContentFiles + Path.DirectorySeparatorChar };
+            // if user specified a PackagePath, then use that. Look for any ** which are indicated by the RecrusiveDir metadata in msbuild.
+            if (customMetadata.Contains("PackagePath"))
+            {
+                targetPaths = packageFile.GetMetadata("PackagePath").Split(';');
+                var recursiveDir = packageFile.GetMetadata("RecursiveDir");
+                if (!string.IsNullOrEmpty(recursiveDir))
+                {
+                    var newTargetPaths = new string[targetPaths.Length];
+                    var i = 0;
+                    foreach (var targetPath in targetPaths)
+                    {
+                        if (targetPath.EndsWith(Path.DirectorySeparatorChar.ToString()))
+                        {
+                            newTargetPaths[i] = Path.Combine(targetPath, recursiveDir);
+                        }
+                        else
+                        {
+                            newTargetPaths[i] = targetPath;
+                        }
+                        
+                        i++;
+                    }
+                    targetPaths = newTargetPaths;
+                }
+            }
+            // this else if condition means the file is within the project directory and the target path should preserve this relative directory structure.
+            else if (sourcePath.StartsWith(currentProjectDirectory, StringComparison.CurrentCultureIgnoreCase) && 
```

Comment:
`sourcePath` is in a [canonical form](https://github.com/NuGet/NuGet.Client/pull/899/files#diff-89dbfecd7235a4059d7d396dbe49c9aeR335), but is `currentProjectDirectory` also?  Should [this](https://github.com/NuGet/NuGet.Client/pull/899/files#diff-89dbfecd7235a4059d7d396dbe49c9aeR220) use `Path.GetFullPath(...)`?